src/Symfony/Component/Cache/Adapter/TagAwareAdapter.php

```diff
@@ -377,57 +375,53 @@ private function generateItems(iterable $items, array $tagKeys): \Generator
         }
     }
 
-    private function getTagVersions(array $tagsByKey, array &$invalidatedTags = [])
+    private function getTagVersions(array $tagsByKey)
     {
-        $tagVersions = $invalidatedTags;
+        $tagVersions = [];
+        $fetchTagVersions = false;
 
         foreach ($tagsByKey as $tags) {
             $tagVersions += $tags;
+
+            foreach ($tags as $tag => $version) {
+                if ($tagVersions[$tag] !== $version) {
+                    unset($this->knownTagVersions[$tag]);
+                }
+            }
         }
 
         if (!$tagVersions) {
             return [];
         }
 
-        if (!$fetchTagVersions = 1 !== \func_num_args()) {
-            foreach ($tagsByKey as $tags) {
-                foreach ($tags as $tag => $version) {
-                    if ($tagVersions[$tag] > $version) {
-                        $tagVersions[$tag] = $version;
-                    }
-                }
-            }
-        }
-
         $now = microtime(true);
         $tags = [];
         foreach ($tagVersions as $tag => $version) {
             $tags[$tag.static::TAGS_PREFIX] = $tag;
-            if ($fetchTagVersions || !isset($this->knownTagVersions[$tag])) {
+            if ($fetchTagVersions || ($this->knownTagVersions[$tag][1] ?? null) !== $version || $now - $this->knownTagVersions[$tag][0] >= $this->knownTagVersionsTtl) {
+                // reuse previously fetched tag versions up to the ttl
                 $fetchTagVersions = true;
-                continue;
-            }
-            $version -= $this->knownTagVersions[$tag][1];
-            if ((0 !== $version && 1 !== $version) || $now - $this->knownTagVersions[$tag][0] >= $this->knownTagVersionsTtl) {
-                // reuse previously fetched tag versions up to the ttl, unless we are storing items or a potential miss arises
-                $fetchTagVersions = true;
-            } else {
-                $this->knownTagVersions[$tag][1] += $version;
             }
         }
 
         if (!$fetchTagVersions) {
             return $tagVersions;
         }
 
+        $newTags = [];
+        $newVersion = null;
         foreach ($this->tags->getItems(array_keys($tags)) as $tag => $version) {
-            $tagVersions[$tag = $tags[$tag]] = $version->get() ?: 0;
-            if (isset($invalidatedTags[$tag])) {
-                $invalidatedTags[$tag] = $version->set(++$tagVersions[$tag]);
+            if (!$version->isHit()) {
+                $newTags[] = $version->set($newVersion ?? $newVersion = random_int(\PHP_INT_MIN, \PHP_INT_MAX));
```

Comment:
`mt_rand()` is faster than `random_int()` and gives the same level of randomness

Reply:
I'd prefer keeping random_int(): the speed difference is negligible compared to network round-trips, and it cannot be seeded, while mt_rand() can, with bad side effects for our use case

Reply:
Agree.